backend/src/auth_service/user_handles/persistence.rs

```diff
@@ -0,0 +1,234 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+use sqlx::{PgExecutor, PgPool, query, query_scalar};
+
+use super::*;
+
+pub(crate) struct UserHandleRecord {
+    pub(crate) user_handle_hash: UserHandleHash,
+    pub(crate) verifying_key: HandleVerifyingKey,
+    pub(crate) expiration_data: ExpirationData,
+}
+
+impl UserHandleRecord {
+    pub(crate) async fn store(&self, executor: impl PgExecutor<'_>) -> sqlx::Result<bool> {
+        let res = query!(
+            "INSERT INTO as_user_handles (
+                hash,
+                verifying_key,
+                expiration_data
+            ) VALUES ($1, $2, $3)
+            ON CONFLICT DO NOTHING",
+            self.user_handle_hash.as_bytes(),
+            self.verifying_key as _,
+            self.expiration_data as _,
+        )
+        .execute(executor)
+        .await?;
+        let inserted = res.rows_affected() > 0;
+        Ok(inserted)
+    }
+
+    pub(crate) async fn load_verifying_key(
+        executor: impl PgExecutor<'_>,
+        hash: &UserHandleHash,
+    ) -> sqlx::Result<Option<HandleVerifyingKey>> {
+        query_scalar!(
+            r#"SELECT verifying_key AS "verifying_key: HandleVerifyingKey"
+                FROM as_user_handles WHERE hash = $1"#,
+            hash.as_bytes(),
+        )
+        .fetch_optional(executor)
+        .await
+    }
+
+    pub(super) async fn delete(
+        executor: impl PgExecutor<'_>,
+        hash: UserHandleHash,
+    ) -> sqlx::Result<bool> {
+        let res = query!(
+            "DELETE FROM as_user_handles WHERE hash = $1",
+            hash.as_bytes(),
+        )
+        .execute(executor)
+        .await?;
+        let deleted = res.rows_affected() > 0;
+        Ok(deleted)
+    }
+
+    #[allow(dead_code)]
+    async fn load_expiration_data(
+        executor: impl PgExecutor<'_>,
+        hash: UserHandleHash,
+    ) -> sqlx::Result<Option<ExpirationData>> {
+        query_scalar!(
+            r#"SELECT expiration_data AS "expiration_data: ExpirationData"
+            FROM as_user_handles WHERE hash = $1"#,
+            hash.as_bytes(),
+        )
+        .fetch_optional(executor)
+        .await
+    }
+
+    pub(super) async fn update_expiration_data(
+        db_pool: &PgPool,
+        hash: UserHandleHash,
+        expiration_data: ExpirationData,
+    ) -> sqlx::Result<bool> {
+        let res = query!(
+            "UPDATE as_user_handles SET expiration_data = $1 WHERE hash = $2",
+            expiration_data as _,
+            hash.as_bytes(),
+        )
+        .execute(db_pool)
+        .await?;
+        let updated = res.rows_affected() > 0;
+        Ok(updated)
+    }
```

Comment:
This would refresh already expired handles, wouldn't it? In that case we should actually delete the handle.